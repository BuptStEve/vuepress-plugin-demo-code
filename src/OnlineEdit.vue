<template>
    <form
        class="demo-and-code-online-edit-wrapper"
        target="_blank"
        :action="actionUrl"
        method="post"
    >
        <!-- https://blog.codepen.io/documentation/api/prefill/ -->
        <template v-if="platform === 'codepen'">
            <input type="hidden" name="data" :value="codepenValue" />
        </template>

        <!-- https://docs.jsfiddle.net/api/display-a-fiddle-from-post -->
        <template v-if="platform === 'jsfiddle'">
            <input type="hidden" name="js" :value="js" />
            <input type="hidden" name="css" :value="css" />
            <input type="hidden" name="html" :value="html" />
            <input type="hidden" name="panel_js" value="3" />
            <input type="hidden" name="resources" :value="resources" />
        </template>

        <!-- https://codesandbox.io/docs/importing#define-api -->
        <template v-if="platform === 'codesandbox'">
            <input type="hidden" name="parameters" :value="codeSandboxValue" />
            <input v-if="codesandboxOptions.query" type="hidden" name="query" :value="codesandboxOptions.query" />
            <input v-if="codesandboxOptions.embed" type="hidden" name="embed" :value="codesandboxOptions.embed" />
            <input v-if="codesandboxOptions.json" type="hidden" name="json" :value="codesandboxOptions.json" />
        </template>

        <button type="submit" :data-tip="platformTip">
            <component :is="platform" />
        </button>
    </form>
</template>

<script>
import codepen from './icons/CodepenIcon.vue'
import jsfiddle from './icons/JsfiddleIcon.vue'
import codesandbox from './icons/CodeSandboxIcon.vue'
import {
    getJsTmpl,
    getHtmlTmpl,
    getCodeSandboxTmpl,
} from './utils'
import {
    PLATFORMS,
    ACTION_MAP,
    PLATFORM_TIP_MAP,
} from './constants'

const vueJs = 'https://unpkg.com/vue/dist/vue.js'

export default {
    name: 'OnlineEdit',
    components: {
        codepen,
        jsfiddle,
        codesandbox,
    },
    props: {
        platform: {
            type: String,
            required: true,
            validator: val => PLATFORMS.indexOf(val) !== -1,
        },
        js: { type: String, default: '' },
        css: { type: String, default: '' },
        html: { type: String, default: '' },
        jsPre: { type: String, default: 'babel' },
        layout: { type: String, default: 'left' },
        jsLibs: { type: Array, default: () => [] },
        cssLibs: { type: Array, default: () => [] },
        editors: { type: String, default: '101' },
        codesandboxOptions: { type: Object, default: () => ({}) },
    },
    computed: {
        jsTmpl: vm => getJsTmpl(vm.js),
        htmlTmpl: vm => getHtmlTmpl(vm.html),
        actionUrl: vm => ACTION_MAP[vm.platform],
        resources: vm => vm.jsLibsWithVue.concat(vm.cssLibs).join(','),
        js_external: vm => vm.jsLibsWithVue.join(';'),
        platformTip: vm => PLATFORM_TIP_MAP[vm.platform],
        css_external: vm => vm.cssLibs.join(';'),
        jsLibsWithVue: vm => vm.jsLibs.concat(vueJs),
        codepenValue: (vm) => JSON.stringify({
            js: vm.jsTmpl,
            css: vm.css,
            html: vm.htmlTmpl,
            layout: vm.layout,
            editors: vm.editors,
            js_external: vm.js_external,
            css_external: vm.css_external,
            js_pre_processor: vm.jsPre,
        }),
        codeSandboxValue: (vm) => getCodeSandboxTmpl({
            js: vm.js,
            css: vm.css,
            html: vm.html,
            deps: vm.codesandboxOptions.deps,
            jsLibs: vm.jsLibs,
            cssLibs: vm.cssLibs,
        }),
    },
}
</script>

<style lang="stylus">
.demo-and-code-online-edit-wrapper {
    display: inline-block;

    vertical-align: middle;

    button {
        position: relative;

        cursor: pointer;

        border-color: transparent;
        outline: none;
        background-color: transparent;

        &:before,
        &:after {
            position: absolute;
            left: 50%;

            transform: translateX(-50%);
        }

        &:hover {
            &:before {
                top: -30px;

                padding: 6px;

                content: attr(data-tip);

                color: #fff;
                border-radius: 4px;
                background-color: rgba(0, 0, 0, .8);

                line-height: 1;
            }

            &:after {
                top: -7px;

                content: '';

                border: 5px solid transparent;
                border-top-color: rgba(0, 0, 0, .8);
            }
        }
    }

    svg {
        width: 34px;
        height: 20px;
    }
}
</style>
