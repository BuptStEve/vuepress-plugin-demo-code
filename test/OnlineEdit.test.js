
import { shallowMount } from '@vue/test-utils'

import OnlineEdit from '@/OnlineEdit.vue'

describe('OnlineEdit', () => {
    let wrapper

    beforeEach(() => {
        wrapper = shallowMount(OnlineEdit, {
            props: {
                platform: 'codepen',
            },
        })
    })

    it('should be rendered', () => {
        expect(wrapper.element).toMatchSnapshot()
    })
})
