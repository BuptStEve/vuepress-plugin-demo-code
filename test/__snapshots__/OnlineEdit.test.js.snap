// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OnlineEdit should be rendered 1`] = `
<form
  action="https://codepen.io/pen/define"
  class="demo-and-code-online-edit-wrapper"
  method="post"
  target="_blank"
>
  <!-- https://blog.codepen.io/documentation/api/prefill/ -->
  <input
    name="data"
    type="hidden"
    value="{\\"js\\":\\"new Vue({\\\\n\\\\tel: '#app', \\\\n\\\\t\\\\n})\\",\\"css\\":\\"\\",\\"html\\":\\"<div id=\\\\\\"app\\\\\\">\\\\n\\\\n\\\\n\\\\n</div>\\",\\"layout\\":\\"left\\",\\"editors\\":\\"101\\",\\"js_external\\":\\"https://unpkg.com/vue@2.6.14/dist/vue.js\\",\\"css_external\\":\\"\\",\\"js_pre_processor\\":\\"babel\\"}"
  />
  <!-- https://docs.jsfiddle.net/api/display-a-fiddle-from-post -->
  <!--v-if-->
  <!-- https://codesandbox.io/docs/importing#define-api -->
  <!--v-if-->
  <button
    data-tip="Codepen"
    type="submit"
  >
    <codepen-stub />
  </button>
</form>
`;
